File: languages/mergesort.py
def merge(a , L , mid , R):
  n1 = mid - L + 1
  n2 = R - mid
  left = [0]*n1
  right = [0]*n2
  for i in range(0 , n1):
    left[i] = a[i + L]
  for i in range(0 , n2):
    right[i] = a[i + mid + 1]
  i =  0; j = 0; k = L
  while(i < n1 and j < n2):
    if(left[i] <= right[j]):
      a[k] = left[i]
      k += 1
      i += 1
    else:
      a[k] = right[j]
      k += 1
      j += 1
  while(i < n1):
    a[k] = left[i]
    k += 1
    i += 1
  while(j < n2):
    a[k] = right[j]
    k += 1
    j += 1

def mergesort(a , L , R):
  if(L < R):
    mid = ( L + R) >> 1;
    mergesort(a , L , mid)
    mergesort(a , mid + 1 , R);
    merge(a , L , mid , R)
     
if __name__ == "__main__":
  a = [9, 1, 7, 4, 6, 2, 8, 3]
  mergesort(a , 0 , 7) 
  assert a == [1, 2, 3, 4, 6, 7, 8, 9], "mergesort failed"

File: languages/mergesort.cc
#include <bits/stdc++.h>
void merge(std :: vector<int>& a , int L , int mid , int R){
  int n1 = mid - L + 1;
  int n2 = R - mid;
  std :: vector<int> left(n1) , right(n2) , res;

  for(int i  =0; i < n1; ++i){
    left[i] = a[i + L];
  }
  for(int i = 0; i < n2; ++i){
    right[i] = a[i + mid + 1];
  }
  int i = 0 , j = 0;
  while(i < n1 && j < n2){
    if(left[i] <= right[j]){
      res.push_back(left[i++]);
    }
    else res.push_back(right[j++]);
  }
  while(i < n1){
    res.push_back(left[i++]);
  }
  while(j < n2){
    res.push_back(right[j++]);
  }
  int k = L;
  for(auto& x: res){
    a[k++] = x;
  }
}
void  mergesort(std:: vector<int>&a , int L , int R){
  if(L < R){
    int mid = (L + R) >> 1;
    mergesort(a , L , mid);
    mergesort(a , mid + 1, R);
    merge(a , L , mid , R);
  }
}
std:: vector<int> fun(std:: vector<int>& a){
  mergesort(a ,0 , a.size() - 1);
  return a;
}
int main(){
  std:: vector<int>a = {9, 1, 7, 4, 6, 2, 8, 3};
  a = fun(a);

  std:: vector<int>expected_result = {1, 2, 3, 4, 6, 7, 8, 9};
  assert (a == expected_result && "Mergesort failed");

}


